src/MagnumPlugins/BasisImporter/BasisImporter.cpp

```diff
@@ -0,0 +1,217 @@
+
+void BasisImporter::doClose() { _in = nullptr; }
+
+void BasisImporter::doOpenData(const Containers::ArrayView<const char> data) {
+    /* Because here we're copying the data and using the _in to check if file
+       is opened, having them nullptr would mean openData() would fail without
+       any error message. It's not possible to do this check on the importer
+       side, because empty file is valid in some formats (OBJ or glTF). We also
+       can't do the full import here because then doImage2D() would need to
+       copy the imported data instead anyway (and the uncompressed size is much
+       larger). This way it'll also work nicely with a future openMemory(). */
+    if(data.empty()) {
+        Error{} << "Trade::BasisImporter::openData(): the file is empty";
+        return;
+    }
+
+    if(!_state->_transcoder.validate_header(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): invalid header";
+        return;
+    }
+
+    if(!_state->_transcoder.start_transcoding(data.data(), data.size())) {
+        Error() << "Trade::BasisImporter::openData(): bad basis file";
+        return;
+    }
+
+    _in = Containers::Array<unsigned char>(data.size());
+    std::copy(data.begin(), data.end(), _in.begin());
+}
+
+UnsignedInt BasisImporter::doImage2DCount() const {
+    return _state->_transcoder.get_total_images(_in.data(), _in.size());
+}
+
+Containers::Optional<ImageData2D> BasisImporter::doImage2D(UnsignedInt index) {
+    const UnsignedInt level = 0;
+
+    const BasisTranscodingType targetFormat =
+        configuration().value<BasisTranscodingType>("format");
+    const basist::transcoder_texture_format format =
+        getTranscoderTextureFormat(targetFormat);
+    const UnsignedInt bytes_per_block = basis_get_bytes_per_block(format);
+
+    basist::basisu_image_info info;
+    if (!_state->_transcoder.get_image_info(_in.data(), _in.size(), info, index)) {
+        Error{} << "Trade::BasisImporter::doImage2D(): unabled to get image info";
+    }
+    const bool hasAlpha = info.m_alpha_flag;
+
+    UnsignedInt origWidth, origHeight, totalBlocks;
+    if (!_state->_transcoder.get_image_level_desc(_in.data(), _in.size(), index,
+            level, origWidth, origHeight, totalBlocks))
+    {
+        Error{} << "Trade::BasisImporter::doImage2D(): unabled to retrieve mip level description";
+        return Containers::NullOpt;
+    }
+
+    const UnsignedInt requiredSize = totalBlocks*bytes_per_block;
+    Containers::Array<char> dest{Containers::NoInit, requiredSize};
+    const UnsignedInt status = _state->_transcoder.transcode_image_level(
+      _in.data(), _in.size(), index, level, dest.data(), dest.size()/bytes_per_block,
+      static_cast<basist::transcoder_texture_format>(format));
+    if(!status) {
+        Error{} << "Trade::BasisImporter::doImage2D(): transcoding failed";
+        return Containers::NullOpt;
+    }
+
+    /* Always using the default 4-byte alignment */
```

Comment:
This is universally true for all compressed formats I think (no format has one- or three-byte blocks), so the comment is redundant.